markdown.go

```diff
@@ -218,7 +218,8 @@ type parser struct {
 	// Footnotes need to be ordered as well as available to quickly check for
 	// presence. If a ref is also a footnote, it's stored both in refs and here
 	// in notes. Slice is nil if footnotes not enabled.
-	notes []*reference
+	notes       []*reference
+	notesRecord map[string]bool
```

Comment:
I see. If we can not reuse `p.refs`, I'm fine with another map.

Regarding bullet 1, I meant to use `map[string]struct{}` instead of `map[string]bool`. Notice how in `isFootnote` you discard the value in the map: `_, ok := p.notesRecord[string(ref.link)]`. Instead of holding the bool that is discarded, you can hold a zero-size empty struct and thus use the map as a set:

declaration: `notesSet map[string]struct{}`
assignment: `p.notesSet[string(lr.link)] = struct{}{}`
and the check remains unchanged: `_, ok := p.notesRecord[string(ref.link)]`